tests/bsim/bluetooth/host/l2cap/credits_seg_recv/src/main.c

```diff
@@ -0,0 +1,446 @@
+		FAIL("Can't enable Bluetooth (err %d)", err);
+		return;
+	}
+
+	LOG_DBG("Peripheral Bluetooth initialized.");
+	LOG_DBG("Connectable advertising...");
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME_OT, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err) {
+		FAIL("Advertising failed to start (err %d)", err);
+		return;
+	}
+
+	LOG_DBG("Advertising started.");
+	LOG_DBG("Peripheral waiting for connection...");
+	WAIT_FOR_FLAG_SET(is_connected);
+	LOG_DBG("Peripheral Connected.");
+
+	int psm = l2cap_server_register(BT_SECURITY_L1);
+
+	LOG_DBG("Registered server PSM %x", psm);
+
+	LOG_DBG("Peripheral waiting for transfer completion");
+	while (rx_cnt < SDU_NUM) {
+		k_sem_take(&sdu_received, K_FOREVER);
+
+		/* Sleep enough so the peer has time to attempt sending another
+		 * SDU. If it still has credits, it's in its right to do so. If
+		 * it does so before we release the ref below, then allocation
+		 * will fail and the channel will be disconnected.
+		 */
+		k_sleep(K_SECONDS(5));
+		LOG_DBG("release SDU ref");
+		LOG_DBG("Giving credits for new SDU.");
+		bt_l2cap_chan_give_credits(&test_ctx.le_chan.chan, 1);
+	}
+
+	bt_conn_foreach(BT_CONN_TYPE_LE, disconnect_device, NULL);
+	LOG_INF("Total received: %d", rx_cnt);
+
+	ASSERT(rx_cnt == SDU_NUM, "Did not receive expected no of SDUs\n");
+
+	PASS("L2CAP CREDITS Peripheral passed\n");
+}
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	struct bt_le_conn_param *param;
+	struct bt_conn *conn;
+	int err;
+
+	err = bt_le_scan_stop();
+	if (err) {
+		FAIL("Stop LE scan failed (err %d)", err);
+		return;
+	}
+
+	char str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, str, sizeof(str));
+
+	LOG_DBG("Connecting to %s", str);
+
+	param = BT_LE_CONN_PARAM_DEFAULT;
+	err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, param, &conn);
+	if (err) {
+		FAIL("Create conn failed (err %d)", err);
+		return;
+	}
+}
+
+static void connect_peripheral(void)
+{
+	struct bt_le_scan_param scan_param = {
+		.type = BT_LE_SCAN_TYPE_ACTIVE,
+		.options = BT_LE_SCAN_OPT_NONE,
+		.interval = BT_GAP_SCAN_FAST_INTERVAL,
+		.window = BT_GAP_SCAN_FAST_WINDOW,
+	};
+
+	UNSET_FLAG(is_connected);
+
+	int err = bt_le_scan_start(&scan_param, device_found);
+
+	ASSERT(!err, "Scanning failed to start (err %d)\n", err);
+
+	LOG_DBG("Central initiating connection...");
+	WAIT_FOR_FLAG_SET(is_connected);
+}
+
+static void connect_l2cap_channel(struct bt_conn *conn, void *data)
+{
+	int err;
+	struct bt_l2cap_le_chan *le_chan = &test_ctx.le_chan;
+
+	*le_chan = (struct bt_l2cap_le_chan){
+		.chan.ops = &ops,
+		.rx.mtu = L2CAP_MTU,
+		.rx.mps = BT_L2CAP_RX_MTU,
+	};
+
+	UNSET_FLAG(flag_l2cap_connected);
+
+	/* Credits can be given before requesting the connection and
+	 * after the 'connected' event. Credits given before connecting
+	 * are sent in the 'initial credits' field of the connection
+	 * request PDU.
+	 */
+	bt_l2cap_chan_give_credits(&le_chan->chan, 1);
```

Comment:
I think the test would still work if you put `2` in there, since `SDU_NUM == 3`.
I'd give slightly better coverage.